## Data Import and glm functions in python:

## 1. Import DebTrivedi dataset and run Poisson Regression from statsmodels

In [1]:
import numpy as np

In [2]:
import pandas as pd

To convert .rda to .csv in R:

    load("~/AnacondaProjects/STAT689/Environments/Classnotes/DebTrivedi.rda")
    write.table(DebTrivedi, file = "DebTrivedi.csv", sep = ",", col.names = NA)


In [ ]:
!cat DebTrivedi.csv

In [3]:
df = pd.read_csv('DebTrivedi.csv',index_col = [0])

In [34]:
df

ofp  ofnp  opp  opnp  emer  hosp     health  numchron adldiff   region  \
1       5     0    0     0     0     1    average         2      no    other   
2       1     0    2     0     2     0    average         2      no    other   
3      13     0    0     0     3     3       poor         4     yes    other   
4      16     0    5     0     1     1       poor         2     yes    other   
5       3     0    0     0     0     0    average         2     yes    other   
6      17     0    0     0     0     0       poor         5     yes    other   
7       9     0    0     0     0     0    average         0      no  midwest   
8       3     0    0     0     0     0    average         0      no  midwest   
9       1     0    0     0     0     0    average         0      no  midwest   
10      0     0    0     0     0     0    average         0      no  midwest   
11      0     0    0     0     0     0    average         1      no  midwest   
12     44     5    2     0     0     1    average         5     yes  midwest   
13      2     0    0     0     0     0    average         1     yes  midwest   
14      1     0    0     0     0     0       poor         1     yes  midwest   
15     19     0    1     2     0     1    average         1      no  midwest   
16     19     0    0     0     0     0    average         0      no  midwest   
17      0     0    0     0     0     0    average         1     yes  midwest   
18      3     9    1     0     0     0    average         2     yes  midwest   
19      2     0    0     0     0     0    average         3      no  midwest   
20     12     0    2     0     0     0    average         4      no  midwest   
21      2     0    1     0     1     2    average         1      no  midwest   
22      3     0    1     0     0     0    average         2      no  midwest   
23      1     0    0     0     0     0    average         1      no  midwest   
24      1     0    0     0     0     0    average         2      no  midwest   
25      5     7    0     0     0     0    average         3      no  midwest   
26      1     0    0     0     0     0    average         1      no  midwest   
27     12     0    0     0     1     0       poor         5      no  midwest   
28      6     0    0     0     0     1       poor         2      no  midwest   
29      2     0    0     0     0     0    average         0      no  midwest   
30      4     0    0     0     0     0    average         3     yes  midwest   
...   ...   ...  ...   ...   ...   ...        ...       ...     ...      ...   
4377   11     0    0     0     0     0    average         1      no    other   
4378   10     0    0     1     0     0    average         0      no    other   
4379    1     0    0     0     0     0    average         3      no    other   
4380    1     0    0     0     0     0    average         1      no    other   
4381    0     0    0     0     0     0    average         0      no    other   
4382    1     0    0     0     0     0  excellent         0      no    other   
4383    2     0    0     0     1     1    average         2      no    other   
4384    3     0    0     0     0     0    average         0      no    other   
4385   10     0    0     0     0     1    average         1      no    other   
4386    7     0    0     0     0     0    average         1      no    other   
4387    0     0    0     0     0     0    average         1      no    other   
4388    1     1    1     0     0     0    average         1      no    other   
4389   16     0    1     0     0     0    average         2      no    other   
4390    6     0    0     0     0     0    average         1      no    other   
4391    3     0    0     0     0     1  excellent         2      no    other   
4392    8     0    0     0     0     0    average         0      no    other   
4393   11     0    0     1     0     0    average         1      no    other   
4394    3     0    0     0     0     0    average         1      no    other   
4395  

In [4]:
sel = np.array([1, 6, 7, 8, 13, 15, 18])-1

In [5]:
df = df.iloc[:,sel]

In [6]:
import statsmodels.api as sm

/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [51]:
#import statsmodels.formula.api as smf

In [8]:
df.head()

ofp  hosp   health  numchron  gender  school privins
1    5     1  average         2    male       6     yes
2    1     0  average         2  female      10     yes
3   13     3     poor         4  female      10      no
4   16     1     poor         2    male       3     yes
5    3     0  average         2  female       6     yes

In [7]:
import patsy
# produce design matrices from R-style formula
formula = 'ofp ~ hosp + health + numchron + gender + school + privins'

In [8]:
y, X = patsy.dmatrices(formula, df, return_type='dataframe')

In [12]:
X

Intercept  health[T.excellent]  health[T.poor]  gender[T.male]  \
1           1.0                  0.0             0.0             1.0   
2           1.0                  0.0             0.0             0.0   
3           1.0                  0.0             1.0             0.0   
4           1.0                  0.0             1.0             1.0   
5           1.0                  0.0             0.0             0.0   
6           1.0                  0.0             1.0             0.0   
7           1.0                  0.0             0.0             0.0   
8           1.0                  0.0             0.0             0.0   
9           1.0                  0.0             0.0             0.0   
10          1.0                  0.0             0.0             0.0   
11          1.0                  0.0             0.0             1.0   
12          1.0                  0.0             0.0             0.0   
13          1.0                  0.0             0.0             0.0   
14          1.0                  0.0             1.0             0.0   
15          1.0                  0.0             0.0             0.0   
16          1.0                  0.0             0.0             0.0   
17          1.0                  0.0             0.0             1.0   
18          1.0                  0.0             0.0             0.0   
19          1.0                  0.0             0.0             1.0   
20          1.0                  0.0             0.0             1.0   
21          1.0                  0.0             0.0             1.0   
22          1.0                  0.0             0.0             1.0   
23          1.0                  0.0             0.0             0.0   
24          1.0                  0.0             0.0             0.0   
25          1.0                  0.0             0.0             0.0   
26          1.0                  0.0             0.0             1.0   
27          1.0                  0.0             1.0             1.0   
28          1.0                  0.0             1.0             0.0   
29          1.0                  0.0             0.0             0.0   
30          1.0                  0.0             0.0             0.0   
...         ...                  ...             ...             ...   
4377        1.0                  0.0             0.0             1.0   
4378        1.0                  0.0             0.0             0.0   
4379        1.0                  0.0             0.0             1.0   
4380        1.0                  0.0             0.0             1.0   
4381        1.0                  0.0             0.0             0.0   
4382        1.0                  1.0             0.0             0.0   
4383        1.0                  0.0             0.0             1.0   
4384        1.0                  0.0             0.0             1.0   
4385        1.0                  0.0             0.0             0.0   
4386        1.0                  0.0             0.0             0.0   
4387        1.0                  0.0             0.0             1.0   
4388        1.0                  0.0             0.0             0.0   
4389        1.0                  0.0             0.0             1.0   
4390        1.0                  0.0             0.0             0.0   
4391        1.0                  1.0             0.0             1.0   
4392        1.0                  0.0             0.0             0.0   
4393        1.0                  0.0             0.0             1.0   
4394        1.0                  0.0             0.0             0.0   
4395        1.0                  1.0             0.0             0.0   
4396        1.0                  0.0             1.0             0.0   
4397        1.0                  0.0             0.0             1.0   
4398        1.0                  0.0             0.0             0.0   
4399        1.0                  0.0             0.0             1.0   
4400        1.0                  0.0          

In [73]:
poisson_mod = sm.GLM(endog = y, exog = X, family = sm.families.Poisson())

In [74]:
poisson_results = poisson_mod.fit()

In [99]:
print(poisson_results.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                    ofp   No. Observations:                 4406
Model:                            GLM   Df Residuals:                     4398
Model Family:                 Poisson   Df Model:                            7
Link Function:                    log   Scale:                             1.0
Method:                          IRLS   Log-Likelihood:                -17972.
Date:                Wed, 28 Mar 2018   Deviance:                       23168.
Time:                        07:54:54   Pearson chi2:                 2.95e+04
No. Iterations:                     5                                         
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               1.0289      0.024     43.257      0.000       0.982       1.075
health[T.excellent]    -0

## 2. Comparison with R output
    fm_pois <- glm(ofp ~ ., data = dt, family = poisson)
    summary(fm_pois)


    Call:
    glm(formula = ofp ~ ., family = poisson, data = dt)

    Deviance Residuals: 
        Min       1Q   Median       3Q      Max  
    -8.4055  -1.9962  -0.6737   0.7049  16.3620  

    Coefficients:
                     Estimate Std. Error z value Pr(>|z|)    
    (Intercept)      1.028874   0.023785  43.258   <2e-16 xxx
    hosp             0.164797   0.005997  27.478   <2e-16 xxx
    healthpoor       0.248307   0.017845  13.915   <2e-16 xxx
    healthexcellent -0.361993   0.030304 -11.945   <2e-16 xxx
    numchron         0.146639   0.004580  32.020   <2e-16 xxx
    gendermale      -0.112320   0.012945  -8.677   <2e-16 xxx
    school           0.026143   0.001843  14.182   <2e-16 xxx
    privinsyes       0.201687   0.016860  11.963   <2e-16 xxx
    
    Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
    
    (Dispersion parameter for poisson family taken to be 1)

    Null deviance: 26943  on 4405  degrees of freedom
    Residual deviance: 23168  on 4398  degrees of freedom
    AIC: 35959

    Number of Fisher Scoring iterations: 5

In [84]:
?sm.GLM()

In [104]:
?sm.discrete.discrete_model.Logit()

Object `sm.discrete.discrete_model.Logit` not found.


In [82]:
?sm.genmod.families.family.Poisson
# The default link for the Poisson family is the log link. 
# Available links are log, identity, and sqrt. 

## 3. Fit Negative Binomial

In [92]:
?sm.genmod.families.family.NegativeBinomial
# The default link for the negative binomial family is the log link.
# Available links are log, cloglog, identity, nbinom and power.

In [98]:
negBinom_mod = sm.GLM(endog = y, exog = X, family = sm.families.NegativeBinomial())
negBinom_results = negBinom_mod.fit()
print(negBinom_results.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                    ofp   No. Observations:                 4406
Model:                            GLM   Df Residuals:                     4398
Model Family:        NegativeBinomial   Df Model:                            7
Link Function:                    log   Scale:                   1.08490597201
Method:                          IRLS   Log-Likelihood:                -12193.
Date:                Wed, 28 Mar 2018   Deviance:                       4455.2
Time:                        07:54:24   Pearson chi2:                 4.77e+03
No. Iterations:                     7                                         
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               0.9260      0.061     15.106      0.000       0.806       1.046
health[T.excellent]    -0

## 4. Comparison with R output:
Notice that the results are different from those in R (The default link in R is also log):  

    dt <- DebTrivedi[, c(1, 6:8, 13, 15, 18)]  
    fm_nbin <- MASS::glm.nb(ofp ~ ., data = dt)  
    summary(fm_nbin)  

    Call:  
    MASS::glm.nb(formula = ofp ~ ., data = dt, init.theta = 1.206603534,  
        link = log)

    Deviance Residuals: 
        Min       1Q   Median       3Q      Max   
    -3.0469  -0.9955  -0.2948   0.2961   5.8185   

    Coefficients:  
                     Estimate Std. Error z value Pr(>|z|)     
    (Intercept)      0.929257   0.054591  17.022  < 2e-16 xxx  
    hosp             0.217772   0.020176  10.793  < 2e-16 xxx  
    healthpoor       0.305013   0.048511   6.288 3.23e-10 xxx  
    healthexcellent -0.341807   0.060924  -5.610 2.02e-08 xxx  
    numchron         0.174916   0.012092  14.466  < 2e-16 xxx  
    gendermale      -0.126488   0.031216  -4.052 5.08e-05 xxx  
    school           0.026815   0.004394   6.103 1.04e-09 xxx  
    privinsyes       0.224402   0.039464   5.686 1.30e-08 xxx  

    Signif. codes:  0 ‘xxx’ 0.001 ‘xx’ 0.01 ‘x’ 0.05 ‘.’ 0.1 ‘ ’ 1  

    (Dispersion parameter for Negative Binomial(1.2066) family taken to be 1)  

    Null deviance: 5743.7  on 4405  degrees of freedom  
    Residual deviance: 5044.5  on 4398  degrees of freedom  
    AIC: 24359  

    Number of Fisher Scoring iterations: 1  


              Theta:  1.2066   
          Std. Err.:  0.0336   

    2 x log-likelihood:  -24341.1070 '  
    
The difference is due to following reason:  
"The Negative Binomial distribution has two parameter, besides the mean parameters there is also a dispersion parameter. GLM cannot estimate it in the model, because it implements only a one-parameter linear exponential family. The user needs to provide this parameter, and it's possible to get the MLE in an extra loop outside of GLM.
discrete.NegativeBinomial estimates the dispersion parameter jointly with the mean parameters by MLE. In econometrics two parameterization for the dispersion function and has exponential distribution as special case. Only one of the parameterization corresponds to the GLM parameterization."  
This is discussed here:
https://github.com/statsmodels/statsmodels/issues/3339

In [9]:
negBinom_mod2 = sm.NegativeBinomial(endog = y, exog = X)
negBinom2_results = negBinom_mod2.fit()
print(negBinom2_results.summary())

Optimization terminated successfully.
         Current function value: 2.762268
         Iterations: 23
         Function evaluations: 26
         Gradient evaluations: 26
                     NegativeBinomial Regression Results                      
Dep. Variable:                    ofp   No. Observations:                 4406
Model:               NegativeBinomial   Df Residuals:                     4398
Method:                           MLE   Df Model:                            7
Date:                Tue, 03 Apr 2018   Pseudo R-squ.:                 0.02580
Time:                        18:48:56   Log-Likelihood:                -12171.
converged:                       True   LL-Null:                       -12493.
                                        LLR p-value:                6.163e-135
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept           

Ideally we'd like to use similar functions for NB and Poisson. Comparing results with sm.Poisson() for Poisson regression:

In [12]:
negBinom2_results.alpha

AttributeError: 'NegativeBinomialResults' object has no attribute 'alpha'

In [107]:
poisson2_mod = sm.Poisson(endog = y, exog = X)
poisson2_results = poisson2_mod.fit()
print(poisson2_results.summary())

Optimization terminated successfully.
         Current function value: 4.078895
         Iterations 23
                          Poisson Regression Results                          
Dep. Variable:                    ofp   No. Observations:                 4406
Model:                        Poisson   Df Residuals:                     4398
Method:                           MLE   Df Model:                            7
Date:                Wed, 28 Mar 2018   Pseudo R-squ.:                 0.09505
Time:                        09:00:00   Log-Likelihood:                -17972.
converged:                       True   LL-Null:                       -19859.
                                        LLR p-value:                     0.000
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               1.0289      0.024     43.257      0.000       0.982       1.075
h

The results are similar.  
##### Final decision: Choose sm.NegetiveBinomial() and sm.Poisson() for starting values.

## 5. Changing optimization parameters
By using `fit(method='bfgs', maxiter=500)` we can choose bfgs instead of Newton-Raphson (default). This will help in setting control parameters. Ideally we'd want same structure for our .comtrol() method as in `fit()` for changing optimization parameters.

In [ ]:
Y = y
## convenience variables
  n <- length(Y)
  kx <- NCOL(X)
  kz <- NCOL(Z)
  Y0 <- Y <= 0
  Y1 <- Y > 0

In [109]:
?sm.discrete.count_model()

Object `sm.discrete.count_model` not found.


In [ ]:
def ziPoisson(parms):
    mu = np.exp(X*)
    
    
    return loglik